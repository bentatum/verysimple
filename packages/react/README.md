# @verysimple/react

This is the `@verysimple/react` package