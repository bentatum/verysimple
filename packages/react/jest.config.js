/** @type {import('jest').Config} */
module.exports = {
  testEnvironment: "jsdom",
  setupFilesAfterEnv: ["@testing-library/jest-dom/extend-expect"],
  collectCoverage: true,
  coverageThreshold: {
    global: {
      statements: 75,
      branches: 72,
      functions: 75,
      lines: 75,
    },
  },
  coverageReporters: ["text-summary", "json-summary"],
  collectCoverageFrom: ["src/**/*.{ts,tsx}", "!**/*.d.ts"],
  coverageDirectory: "coverage",
  moduleNameMapper: {
    "@/test/(.*)": "<rootDir>/test/$1",
    "@/(.*)$": "<rootDir>/src/$1",
  },
  modulePathIgnorePatterns: ["<rootDir>/dist/"],
  prettierPath: null,
};
