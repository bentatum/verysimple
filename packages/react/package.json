{
  "name": "@verysimple/react",
  "version": "8.1.1",
  "description": "A very simple component library",
  "main": "dist/index.js",
  "files": [
    "dist"
  ],
  "author": "Ben Tatum <mail@benjtatum.com>",
  "license": "MIT",
  "repository": "https://github.com/bentatum/verysimple",
  "scripts": {
    "test": "jest",
    "prebuild": "rm -rf dist",
    "build": "tsc --project tsconfig.build.json && tsc-alias"
  },
  "peerDependencies": {
    "react": ">=16",
    "react-dom": ">=16"
  },
  "devDependencies": {
    "@babel/core": "^7.20.12",
    "@babel/preset-env": "^7.19.1",
    "@babel/preset-react": "^7.18.6",
    "@babel/preset-typescript": "^7.18.6",
    "@testing-library/dom": "^8.20.0",
    "@testing-library/jest-dom": "^5.16.5",
    "@testing-library/react": "^13.4.0",
    "@testing-library/user-event": "^14.4.3",
    "@types/inquirer": "latest",
    "@types/jest": "^29.2.3",
    "@types/react": "^18.0.21",
    "@types/testing-library__jest-dom": "^5.14.5",
    "babel-jest": "^29.0.3",
    "babel-loader": "^8.2.5",
    "classnames": "^2.3.2",
    "framer-motion": "^10.15.0",
    "jest": "^29.0.3",
    "jest-environment-jsdom": "^29.0.3",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "tailwindcss": "^3.1.8",
    "tsc-alias": "^1.8.2",
    "typescript": "^4.9.3"
  },
  "volta": {
    "node": "20.10.0",
    "yarn": "1.22.21"
  },
  "dependencies": {
    "@headlessui/react": "^2.1.9",
    "@heroicons/react": "^2.0.13",
    "clsx": "^2.1.0",
    "framer-motion": "^10.15.0"
  }
}
