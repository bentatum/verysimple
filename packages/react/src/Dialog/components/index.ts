export * from './DialogTitle';
export * from './DialogCloseButton';
