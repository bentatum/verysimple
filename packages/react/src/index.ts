export * from "./ButtonBase";
export * from "./Card";
export * from "./Dialog";
export * from "./FilledButtonBase";
export * from "./FilledIconButton";
export * from "./GradientBorder";
export * from "./helpers";
export * from "./IconButton";
export * from "./NormalIconButton";
export * from "./OutlinedButtonBase";
export * from "./OutlinedIconButton";
export * from "./Spinner";
export * from "./types";
