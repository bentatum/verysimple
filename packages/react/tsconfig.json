{
  "extends": "@verysimple/tsconfig/react-library.json",
  "compilerOptions": {
    "outDir": "./dist",
    "paths": {
      "@/*": ["./src/*"]
    }
  },
  "include": ["src/**/*.ts", "src/**/*.tsx"],
  "exclude": ["node_modules"]
}
