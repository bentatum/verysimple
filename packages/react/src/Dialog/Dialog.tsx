import { FC, forwardRef } from "react";
import { Dialog as DialogPrimitive } from "@headlessui/react";
import clsx from "clsx";
import DialogCloseButton from "./components/DialogCloseButton";
import type { $ElementProps } from "../types";
import Card, { CardProps } from "../Card";

export type DialogProps = $ElementProps<typeof DialogPrimitive> &
  CardProps & {
    overlayProps?: $ElementProps<typeof DialogPrimitive.Overlay>;
    showCloseButton?: boolean;
  };

export const Dialog: FC<DialogProps> = forwardRef(
  (
    {
      children,
      open,
      overlayProps,
      onClose,
      showCloseButton,
      initialFocus,
      className = "",
      disableOverlay,
      ...props
    },
    ref
  ) => {
    return (
      <DialogPrimitive
        ref={ref}
        open={open}
        initialFocus={initialFocus}
        {...props}
        onClose={onClose}
        className="fixed z-30 inset-0 overflow-y-auto"
      >
        <div className="flex items-center justify-center min-h-screen">
          <DialogPrimitive.Overlay
            {...overlayProps}
            className={clsx(
              "z-10 fixed inset-0 bg-black/40 backdrop-blur",
              overlayProps?.className
            )}
          />
          <Card
            className={clsx(
              "z-20 shadow-2xl relative",
              {
                "p-7": !className.match(/p-/),
                "m-7": !className.match(/m-/),
                "w-full max-w-prose": !className.match(/w-|max-w-/),
                "bg-foreground": !className.match(/bg-/),
              },
              className
            )}
          >
            <>
              {showCloseButton && (
                <div className="hidden sm:block absolute top-3 right-3 z-10">
                  <DialogCloseButton onClick={() => onClose(true)} />
                </div>
              )}
              {children}
            </>
          </Card>
        </div>
      </DialogPrimitive>
    );
  }
);

Dialog.displayName = "Dialog";

export default Dialog;
