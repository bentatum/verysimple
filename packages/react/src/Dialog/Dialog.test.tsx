import { render, screen } from "@testing-library/react";
import Dialog from "./Dialog";
import "@/test/mocks/window/IntersectionObserver";
import userEvent from "@testing-library/user-event";
import { createRef, ReactNode } from "react";
import { Transition } from "@headlessui/react";
import { createMock } from "ts-jest-mock";

jest.mock("@headlessui/react", () => {
  const { Transition, ...rest } = jest.requireActual("@headlessui/react");
  return {
    ...rest,
    Transition: {
      ...Transition,
      render: jest.fn(({ children }) => {
        return <div>{children as ReactNode}</div>;
      }),
      Child: jest.fn(({ children }) => {
        return <div>{children as ReactNode}</div>;
      }),
    },
  };
});
const TransitionMock = createMock((Transition as any).render);

beforeEach(() => {
  // mute focus trap warning
  jest.spyOn(console, "warn").mockImplementation(() => {});
});

describe("Dialog", () => {
  test("default render", () => {
    const { baseElement } = render(
      <Dialog open onClose={() => {}}>
        test
      </Dialog>
    );
    expect(baseElement).toMatchSnapshot();
    expect(screen.getByRole("dialog")).toBeInTheDocument();
    expect(
      screen.queryByRole("button", { name: "Close" })
    ).not.toBeInTheDocument();
  });

  test("closing", async () => {
    const user = userEvent.setup();
    const onClose = jest.fn();
    const { baseElement, rerender } = render(
      <Dialog open onClose={onClose} showCloseButton>
        test
      </Dialog>
    );
    expect(baseElement).toMatchSnapshot();
    await user.click(screen.getByRole("button", { name: "Close" }));
    expect(onClose).toBeCalledTimes(1);
    TransitionMock.mockImplementationOnce(({ children, afterLeave }: any) => {
      afterLeave!();
      return <div>{children as ReactNode}</div>;
    });
    rerender(
      <Dialog open={false} onClose={onClose} showCloseButton>
        test
      </Dialog>
    );
    expect(baseElement).toMatchSnapshot();
    expect(screen.queryByRole("dialog")).not.toBeInTheDocument();
  });

  it("forwards the ref", () => {
    const ref = createRef<HTMLDivElement>();
    render(
      <Dialog ref={ref} open={true} onClose={() => {}}>
        test
      </Dialog>
    );
    expect(ref.current).toBeInTheDocument();
  });

  it("has a display name", () => {
    expect(Dialog.displayName).toBe("Dialog");
  });
});
