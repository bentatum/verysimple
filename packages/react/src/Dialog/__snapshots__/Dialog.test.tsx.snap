// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog closing 1`] = `
<body>
  <div>
    <div
      style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
    />
  </div>
  <div
    id="headlessui-portal-root"
  >
    <div
      data-headlessui-portal=""
    >
      <button
        aria-hidden="true"
        data-headlessui-focus-guard="true"
        style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px;"
        type="button"
      />
      <div>
        <div
          aria-modal="true"
          class="fixed z-30 inset-0 overflow-y-auto"
          data-headlessui-state="open"
          id="headlessui-dialog-:r2:"
          role="dialog"
        >
          <div
            class="flex items-center justify-center min-h-screen"
          >
            <div
              aria-hidden="true"
              class="z-10 fixed inset-0 bg-black/40 backdrop-blur"
              data-headlessui-state="open"
              id="headlessui-dialog-overlay-:r3:"
            />
            <div
              class="rounded z-20 shadow-2xl relative p-7 m-7 w-full max-w-prose bg-foreground"
            >
              <div
                class="hidden sm:block absolute top-3 right-3 z-10"
              >
                <button
                  aria-label="Close"
                  class="focus:outline-none cursor-pointer disabled:cursor-not-allowed inline-flex items-center justify-center min-w-[48px] shadow-sm hover:shadow-md disabled:shadow-none transition-shadow ease-in duration-100 rounded text-base min-h-[44px] focus:bg-gray-300/20 hover:bg-gray-300/20 active:bg-gray-300/20 dark:focus:bg-gray-750/60 dark:hover:bg-gray-750/60 dark:active:bg-gray-750/60"
                  tabindex="0"
                >
                  <svg
                    aria-hidden="true"
                    class="h-7 w-7"
                    data-slot="icon"
                    fill="currentColor"
                    viewBox="0 0 24 24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      clip-rule="evenodd"
                      d="M5.47 5.47a.75.75 0 0 1 1.06 0L12 10.94l5.47-5.47a.75.75 0 1 1 1.06 1.06L13.06 12l5.47 5.47a.75.75 0 1 1-1.06 1.06L12 13.06l-5.47 5.47a.75.75 0 0 1-1.06-1.06L10.94 12 5.47 6.53a.75.75 0 0 1 0-1.06Z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </button>
              </div>
              test
            </div>
          </div>
        </div>
      </div>
      <button
        aria-hidden="true"
        data-headlessui-focus-guard="true"
        style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px;"
        type="button"
      />
    </div>
  </div>
</body>
`;

exports[`Dialog closing 2`] = `
<body>
  <div>
    <div
      style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
    />
  </div>
  <div
    id="headlessui-portal-root"
  >
    <div
      data-headlessui-portal=""
    >
      <div />
    </div>
  </div>
</body>
`;

exports[`Dialog default render 1`] = `
<body>
  <div>
    <div
      style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
    />
  </div>
  <div
    id="headlessui-portal-root"
  >
    <div
      data-headlessui-portal=""
    >
      <button
        aria-hidden="true"
        data-headlessui-focus-guard="true"
        style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px;"
        type="button"
      />
      <div>
        <div
          aria-modal="true"
          class="fixed z-30 inset-0 overflow-y-auto"
          data-headlessui-state="open"
          id="headlessui-dialog-:r0:"
          role="dialog"
        >
          <div
            class="flex items-center justify-center min-h-screen"
          >
            <div
              aria-hidden="true"
              class="z-10 fixed inset-0 bg-black/40 backdrop-blur"
              data-headlessui-state="open"
              id="headlessui-dialog-overlay-:r1:"
            />
            <div
              class="rounded z-20 shadow-2xl relative p-7 m-7 w-full max-w-prose bg-foreground"
            >
              test
            </div>
          </div>
        </div>
      </div>
      <button
        aria-hidden="true"
        data-headlessui-focus-guard="true"
        style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px;"
        type="button"
      />
    </div>
  </div>
</body>
`;
